feat(self-managed): aws opensearch doc (#4406)
camunda/camunda-docs: docs/self-managed/setup/deploy/amazon/amazon-eks/eksctl.md

---
id: eks-eksctl

> title: "Deploy an EKS cluster with eksctl (fast-try)"

Reviewer: not sure about the `(fast-try)` in the title.

Maybe we could utilize tags - https://docusaurus.io/docs/2.x/create-doc#doc-tags
and then maybe something like `quick guide`.

Reviewer: maybe just `quickstart`.

Like with manuals, you usually have the complete manual and the quickstart guide

description: "Deploy an Amazon Kubernetes cluster (EKS) with eksctl with step-by-step guidance."
---

This guide explores the streamlined process of deploying Camunda 8 Self-Managed on Amazon Elastic Kubernetes Service (EKS) using the `eksctl` command-line tool.

[Eksctl](https://eksctl.io/) is a common CLI tool for quickly creating and managing your Amazon EKS clusters and is [officially endorsed](https://docs.aws.amazon.com/eks/latest/userguide/getting-started-eksctl.html) by Amazon.

> While this guide is suitable for testing purposes, building a robust, scalable, and reproducible infrastructure is better achieved using Infrastructure as Code (IaC) tools like describes the [Terraform Guide](./terraform-setup.md), which offers more flexibility and control over the cloud environment.

Reviewer: this still applies

> While this guide is suitable for testing purposes, building a robust, scalable, and reproducible infrastructure is better achieved using Infrastructure as Code (IaC) tools like describes the [Terraform Guide](./terraform-setup.md), which offers more flexibility and control over the cloud environment.

Reviewer: nvm, was in the wrong diff.

---
id: eks-eksctl

> title: "Deploy an EKS cluster with eksctl (quickstart)"

Reviewer 2: I feel like "quickstart" is a little awkward here, as most of our docs could be categorized as "quickstart?" WDYT @conceptualshark?

---

This guide explores the streamlined process of deploying Camunda 8 Self-Managed on Amazon Elastic Kubernetes Service (EKS) using the `eksctl` command-line tool.

[Eksctl](https://eksctl.io/) is a common CLI tool for quickly creating and managing your Amazon EKS clusters and is [officially endorsed](https://docs.aws.amazon.com/eks/latest/userguide/getting-started-eksctl.html) by Amazon.

While this guide is suitable for testing purposes, building a robust, scalable, and reproducible infrastructure is better achieved using Infrastructure as Code (IaC) tools like those described in the [Terraform Guide](./terraform-setup.md), which offers more flexibility and control over your Cloud environment.

This guide provides a user-friendly approach for setting up and managing Amazon EKS clusters. It covers everything from the prerequisites, such as AWS IAM role configuration, to creating a fully functional Amazon EKS cluster and a managed Aurora PostgreSQL instance. Ideal for those seeking a practical and efficient method to deploy Camunda 8 on AWS, this guide provides detailed instructions for setting up the necessary environment and AWS IAM configurations.

> ## 0. Prerequisites

Reviewer 2: If this is meant as an order of steps, I would start with 1. If not, I would remove the numbers entirely.

- A [managed Aurora PostgreSQL 15.x](https://aws.amazon.com/rds/aurora/) instance that will be used by the Camunda 8 components.
- A [managed OpenSearch domain](https://aws.amazon.com/opensearch-service/) created and configured for use with the Camunda platform..
- [IAM Roles for Service Accounts](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html) (IRSA) configured and [Pod Identities](https://docs.aws.amazon.com/eks/latest/userguide/pod-identities.html).
  - This simplifies the setup by not relying on explicit credentials, but instead allows creating a mapping between IAM roles and Kubernetes service accounts based on a trust relationship. A [blog post](https://aws.amazon.com/blogs/containers/diving-into-iam-roles-for-service-accounts/) by AWS visualizes this on a technical level.
  - This allows a Kubernetes service account to temporarily impersonate an AWS IAM role to interact with AWS services like S3, RDS, or Route53 without supplying explicit credentials.

This basic cluster setup is required to continue with the Helm set up as described in our [AWS Helm guide](./eks-helm.md).

#### Variants

> We refer to this architecture as **Standard Installation** ([more informations](./terraform-setup.md#variants)), which can be set up with or without a **Domain** ([ingress](https://docs.aws.amazon.com/eks/latest/userguide/alb-ingress.html)).

Reviewer 2: ```suggestion
We refer to this architecture as [**standard installation**](./terraform-setup.md#variants), which can be set up with or without a **Domain** ([ingress](https://docs.aws.amazon.com/eks/latest/userguide/alb-ingress.html)).
```

:::caution Ownership of the created resources

A user who creates resources in AWS will always retain administrative access to those resources, including any Kubernetes clusters. It is recommended to create a dedicated [AWS IAM user](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_users.html) to ensure that the resources are managed and owned by that specific user.
This ensures that the user maintains admin access to Kubernetes and associated resources unless those resources are explicitly deleted.

[Create access keys](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_access-keys.html) for the new IAM user via the console and export them as `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` variables to use with the AWS CLI and `eksctl`

:::

### Setup eksctl

The [eksctl](https://eksctl.io/) is a tool that allows the creation of clusters via a single command, but this doesn't support all configuration options. Therefore, we're supplying a YAML file that can be used with the CLI to create the cluster preconfigured with various settings.

> This [installation guide](https://eksctl.io/installation/) explains how to install it.

Reviewer 2: ```suggestion
Review the [installation guide](https://eksctl.io/installation/) for additional details.
```

## 1. Provisioning the complete infrastructure for Camunda 8 on AWS

### Set Up AWS authentication

Use the [AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-getting-started.html).

```bash
# set your region
export AWS_REGION="eu-central-1"

aws configure
```

Enter your `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, `AWS_REGION`, and output format which can be retrieved from the [AWS Console](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_access-keys.html).

:::caution Ownership of the created resources

A user who creates resources in AWS will always retain administrative access to those resources, including any Kubernetes clusters created. It is recommended to create a dedicated [AWS IAM user](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_users.html), ensuring that the resources are managed and owned by that user.

[Create access keys](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_access-keys.html) for the new IAM user via the console and export them as `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` variables to use with the AWS CLI and `eksctl`

> This means that the user will always have admin access on Kubernetes unless you delete it.

Reviewer: it's a bit fuzzy what `it` refers to.
It the user?  the cluster? or the admin access?

While the guide is primarily tailored for UNIX systems, it can also be run under Windows by utilizing the [Windows Subsystem for Linux](https://learn.microsoft.com/windows/wsl/about).

> :::warning Cost Management

Reviewer: just for consistency, you used in the other doc `Cost management`, which generally aligns with the titles.
Doesn't matter too much since the admonition puts it anyway in uppercase when rendered.

- [IAM Roles for Service Accounts](https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html) (IRSA) configured.
  - This simplifies the setup by not relying on explicit credentials, but instead allows creating a mapping between IAM roles and Kubernetes service accounts based on a trust relationship. A [blog post](https://aws.amazon.com/blogs/containers/diving-into-iam-roles-for-service-accounts/) by AWS visualizes this on a technical level.
  - This allows a Kubernetes service account to temporarily impersonate an AWS IAM role to interact with AWS services like S3, RDS, or Route53 without supplying explicit credentials.

> - An OpenSearch domain created and configured for use with the Camunda 8 platform, leveraging the capabilities of [AWS OpenSearch Service](https://aws.amazon.com/opensearch-service/).

Reviewer: ```suggestion
- An OpenSearch domain created and configured for use with the Camunda 8 platform, leveraging the capabilities of the [AWS OpenSearch Service](https://aws.amazon.com/opensearch-service/).
```

> - An OpenSearch domain created and configured for use with the Camunda 8 platform, leveraging the capabilities of [AWS OpenSearch Service](https://aws.amazon.com/opensearch-service/).

Reviewer: another topic here, not sure what the docs align on but you've used `Camunda 8 platform` and `Camunda platform`.

> - An OpenSearch domain created and configured for use with the Camunda 8 platform, leveraging the capabilities of [AWS OpenSearch Service](https://aws.amazon.com/opensearch-service/).

Reviewer: also would mention it to be a managed OpenSearch DB

For more information concerning the KMS encryption, refer to the [eksctl documentation](https://eksctl.io/usage/kms-encryption/).

> ### Create the cluster using eksctl

Reviewer: wondering about the yaml below whether it would make sense to put into `camunda-tf-eks-module` as well and reference.

metadata:
  name: ${CLUSTER_NAME:-camunda-cluster} # e.g. camunda-cluster
  region: ${REGION:-eu-central-1} # e.g. eu-central-1

>   version: "1.31"

Reviewer: > Error: invalid version, supported values: 1.23, 1.24, 1.25, 1.26, 1.27, 1.28, 1.29, 1.30

Reviewer: TIL, didn't think it was dependent on the eksctl verison.
I was on 0.190, my bad.

With eksctl you can execute the previously created file as follows and takes 25-30 minutes.

```shell

> cat cluster.yaml

Reviewer: why are we doing `cat cluster.yaml`?

  </p>
</details>

> ## PostgreSQL Database

Reviewer: afaik doc titles are always uppercase followed by lowercase, except it's a proper noun.

Creating a PostgreSQL database can be accomplished through various methods, such as using the AWS Management Console or the AWS CLI. This guide focuses on providing a reproducible setup using the CLI. For information on creating PostgreSQL using the UI, refer to the [AWS documentation](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/CHAP_GettingStarted.CreatingConnecting.PostgreSQL.html).

> The resulting PostgreSQL instance and the default database `camunda` is intended for use with Keycloak. Additional databases can be manually added post-creation for Identity with multi-tenancy, although this guide will not cover that process as the default for multi-tenancy is disabled.

Reviewer: Reminder for myself to double check this. Might be covered somewhere else actually.

> ### Steps to Create a PostgreSQL Database

Reviewer: again afaik, titles are uppercase followed by lowercase except for proper nouns.

       --database-name $DEFAULT_DB_NAME \
       --db-subnet-group-name camunda-postgres
   ```

   - More configuration options can be found in the [AWS documentation](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/rds/create-db-cluster.html).

7. **Wait for the PostgreSQL cluster to be ready:**

   ```shell
   aws rds wait db-cluster-available \
       --db-cluster-identifier $RDS_NAME
   ```

8. **Create a database instance within the DB cluster:**

   Ensure that the `engine-version` matches the previously created PostgreSQL cluster.

   ```shell
   aws rds create-db-instance \
       --db-instance-identifier $RDS_NAME \
       --db-cluster-identifier $RDS_NAME \
       --engine aurora-postgresql \
       --engine-version $POSTGRESQL_VERSION \
       --no-publicly-accessible \
       --db-instance-class db.t3.medium
   ```

   - More configuration options can be found in the [AWS documentation](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/rds/create-db-instance.html).

9. **Wait for changes to be applied:**

   ```shell
   aws rds wait db-instance-available \
       --db-instance-identifier $RDS_NAME
   ```

> ### Verifying Connectivity Between the Amazon EKS Cluster and the PostgreSQL Database

Reviewer: title again

1. **Retrieve the writer endpoint of the DB cluster:**

   ```shell
   export DB_HOST=$(aws rds describe-db-cluster-endpoints \
     --db-cluster-identifier $RDS_NAME \
     --query "DBClusterEndpoints[?EndpointType=='WRITER'].Endpoint" \
     --output text)
   ```

2. **Start an Ubuntu container in interactive mode within the Amazon EKS cluster:**

   ```shell
   kubectl run ubuntu --rm -i --tty --image ubuntu --env="DB_HOST=$DB_HOST" --env="PG_USERNAME=$PG_USERNAME" -- bash
   ```

3. **Install required dependencies:**

   ```shell
   apt update && apt install -y postgresql-client
   ```

4. **Connect to the PostgreSQL database:**

   ```shell
   psql \
     --host=$DB_HOST \
     --username=$PG_USERNAME \
     --port=5432 \
     --dbname=postgres
   ```

   Verify that the connection is successful.

> ## OpenSearch Domain

Reviewer: title

Creating an OpenSearch domain can be accomplished in various ways. This guide provides a reproducible setup using the AWS CLI. For additional information on using the AWS Management Console, refer to the [OpenSearch documentation](https://opensearch.org/docs/latest/index/).

> The resulting OpenSearch domain is intended for use with the Camunda platform.

Reviewer: here is the other occurrence that I mentioned above about the `Camunda 8 platform` vs the `Camunda platform`.

> ### Step-by-Step Setup

Reviewer: title

   ```

   The variable `GROUP_ID` contains the output (the value should look like this: `sg-1234567890`).

3. **Create a security ingress rule to allow access to OpenSearch over HTTPS (port 443) from within the VPC:**

   ```shell
   aws ec2 authorize-security-group-ingress \
     --group-id $GROUP_ID \
     --protocol tcp \
     --port 443 \
     --cidr $CIDR  # Replace with the CIDR range of your EKS cluster, e.g., <EKS_CIDR_BLOCK>
   ```

   Ensure that the CIDR range is appropriate for your environment.

4. **Retrieve the private subnets of the VPC:**

   ```shell
   export SUBNET_IDS=$(aws ec2 describe-subnets \
     --filter Name=vpc-id,Values=$VPC_ID \
     --query "Subnets[?Tags[?Key=='aws:cloudformation:logical-id']|[?contains(Value, 'Private')]].SubnetId" \
     --output text | expand -t 1)
   ```

   The variable `SUBNET_IDS` now contains the output values of the private subnets (the value should look like this: `subnet-0123456789 subnet-1234567890`).

5. **Create the OpenSearch domain:**

   ```shell
   export OPENSEARCH_DOMAIN_NAME=<your-domain-name>  # Replace with your desired domain name

   aws opensearch create-domain --domain-name $OPENSEARCH_DOMAIN_NAME \

>      --elasticsearch-version 2.15 \

Reviewer: I'm surprised this would work at all since I don't see the option `--elasticsearch-version` for either v1 or v2 cli.

always just `--engine-version OpenSearch_2.15`

     --node-to-node-encryption-options Enabled=true \
     --encryption-at-rest-options Enabled=true \

>      --vpc-options SubnetIds=[${SUBNET_IDS}],SecurityGroupIds=[${GROUP_ID}]

Reviewer: I'd suggest adding a default Instance type.
I haven't checked what the command results into but e.g. the new OpenSearch optimized OR1 instances are not compatible.

     # TODO: add password and user
     # TODO: use opensearch, not elastic
   ```

6. **Wait for the OpenSearch domain to be active:**

   ```shell
   aws opensearch wait domain-active --domain-name $OPENSEARCH_DOMAIN_NAME
   ```

7. **Retrieve the endpoint of the OpenSearch domain:**

   ```shell
   export OPENSEARCH_ENDPOINT=$(aws opensearch describe-domains --domain-names $OPENSEARCH_DOMAIN_NAME --query "DomainStatusList[0].Endpoint" --output text)
   ```

   This endpoint will be used to connect to your OpenSearch domain.

> ### Verifying Connectivity from Within the EKS Cluster

Reviewer: title

> To verify that the OpenSearch domain is accessible only from within your Amazon EKS cluster, follow these steps:

Reviewer: I'd rephrase this as to either say since it's only accessible from within cluster bla bla or remove the only.

Since the following does not check whether it's only accessible from within the cluster.

1. **Deploy a temporary pod to test connectivity:**

   Create a temporary pod using a lightweight image, like `curlimages/curl`, to test the connection to OpenSearch.

   ```shell
   kubectl run curl-opensearch --rm -i --tty --image curlimages/curl -- sh
   ```

2. **Within the pod, test connectivity to the OpenSearch domain:**

   Once inside the pod, use the following command to test the connection to the OpenSearch endpoint:

   ```shell
   curl -XGET https://$OPENSEARCH_ENDPOINT
   ```

   If everything is set up correctly, you should receive a response from the OpenSearch service.

> You have successfully set up an OpenSearch domain that is accessible only from within your Amazon EKS cluster. Be sure to customize security group rules and CIDR ranges according to your environment's requirements. For further details, refer to the [OpenSearch documentation](https://opensearch.org/docs/latest/index/).

Reviewer: depending on how you rephrase the above suggestion, adjust the only or leave it depending on the outcome.

  </p>
</details>

> ### Access the created EKS Cluster

Reviewer 2: ```suggestion
### Access the created EKS cluster
```

You can gain access to the Amazon EKS cluster via the `AWS CLI` using the following command:

```shell
aws eks --region "$REGION" update-kubeconfig --name "$CLUSTER_NAME" --alias "$CLUSTER_NAME"
```

After updating the kubeconfig, you can verify your connection to the cluster with kubectl:

```shell
kubectl get nodes
```

> Then ensure that a namespace exists for camunda.

Reviewer 2: ```suggestion
Ensure a namespace exists for Camunda.
```

In the following procedure, we reference the `camunda` namespace to create some required resources in the Kubernetes cluster, such as secrets or one-time setup jobs.

```shell
kubectl create namespace camunda
```

> ### Check Existing StorageClasses

Reviewer 2: ```suggestion
### Check existing StorageClasses
```

Reviewer 2: Will need to ensure sentence case structure across all of these new headers.

To see the available **StorageClasses** in your Kubernetes cluster, including which one is set as default, use the following command:

```bash
kubectl describe storageclass
```

#### StorageClass

To check if `gp3` is set as the default StorageClass, you can look for the annotation `storageclass.kubernetes.io/is-default-class: "true"` in the output of the previous command. If `gp3` is not installed or not set as default StorageClass, you need to install it and set it as default.

We recommend using **gp3** volumes with Camunda 8 (see [volume performance](./amazon-eks.md#volume-performance)).
It may be necessary to create the StorageClass, as the default configuration only includes **gp2**. For detailed information, refer to the [AWS documentation](https://aws.amazon.com/ebs/general-purpose/).

#### Create `gp3` StorageClass

1. **Create the `gp3` StorageClass.**

```shell
cat << EOF | kubectl apply -f -
---
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: ebs-sc
  annotations:
    storageclass.kubernetes.io/is-default-class: "true"
provisioner: ebs.csi.aws.com
parameters:
  type: gp3
reclaimPolicy: Retain
volumeBindingMode: WaitForFirstConsumer
EOF
```

> Please note that the `ebs-sc` StorageClass relies on the `ebs.csi.aws.com` provisioner, which is provided by the **aws-ebs-csi-driver** addon. This addon was installed during the cluster creation.

Reviewer 2: Could convert into an admonition. Also, removal of "please" 👍

For more information, refer to the [official AWS documentation](https://docs.aws.amazon.com/eks/latest/userguide/ebs-csi.html).

2. **Modify the `gp2` StorageClass to mark it as a non-default StorageClass:**

```shell
kubectl patch storageclass gp2 -p '{"metadata": {"annotations":{"storageclass.kubernetes.io/is-default-class":"false"}}}'
```

> After executing these commands, you will have a `gp3` StorageClass set as the default and the `gp2` StorageClass marked as non-default, provided that **gp2** was already present. Ensure to verify the changes again by running the `kubectl get storageclass` command.

Reviewer 2: ```suggestion
After executing these commands, you will have a `gp3` StorageClass set as the default and the `gp2` StorageClass marked as non-default, provided that **gp2** was already present. Verify the changes again by running the `kubectl get storageclass` command.
```

By default, the cluster uses **Pod Identity** to manage IAM roles for your applications. This means that service accounts are associated with IAM roles, allowing your pods to securely access AWS resources without hardcoding credentials. For more information on configuring Pod Identity, refer to the [AWS documentation](https://docs.aws.amazon.com/eks/latest/userguide/pod-identities.html).

Additionally, to [enable OpenID Connect (OIDC) and IAM Roles for Service Accounts (IRSA)](https://docs.aws.amazon.com/eks/latest/userguide/enable-iam-roles-for-service-accounts.html) on your cluster, you can use the following commands:

1. **Determine the OIDC issuer ID for your cluster:**

   First, ensure that your EKS cluster is set up with an OIDC provider. The following command should show you the OIDC issuer:

   ```bash
   export oidc_id=$(aws eks describe-cluster --name "$CLUSTER_NAME" --query "cluster.identity.oidc.issuer" --output text)
   echo "$oidc_id"
   ```

   Determine whether an IAM OIDC provider with your cluster’s issuer ID is already in your account.

   ```bash
   aws iam list-open-id-connect-providers | grep $oidc_id | cut -d "/" -f4
   ```

> 

Reviewer: I was having a problem with this one detecting anything.

`oidc_id=https://oidc.eks.eu-central-1.amazonaws.com/id/3C1A33F675599F7B5AD3C663E2CC2DCE`

```
{
    "OpenIDConnectProviderList": [
        {
            "Arn": "arn:aws:iam::444804106854:oidc-provider/oidc.eks.eu-central-1.amazonaws.com/id/3C1A33F675599F7B5AD3C663E2CC2DCE"
        },
        {
            "Arn": "arn:aws:iam::444804106854:oidc-provider/oidc.eks.eu-west-2.amazonaws.com/id/0184EE512F7947A1EDABDCFCC69F158F"
        },
        {
            "Arn": "arn:aws:iam::444804106854:oidc-provider/oidc.eks.eu-west-2.amazonaws.com/id/4EBF8FD07659D7F8F4BC2DC11F9A282C"
        },
        {
            "Arn": "arn:aws:iam::444804106854:oidc-provider/oidc.eks.eu-west-2.amazonaws.com/id/6B0023DBBADEB00479FF98F82F8CC427"
        },
        {
            "Arn": "arn:aws:iam::444804106854:oidc-provider/oidc.eks.eu-west-2.amazonaws.com/id/B6946796624DBD2C57423647D969FF7F"
        }
    ]
}
```

---

as discussed just `| cut -d '/' -f 5`, see official docs again.

  </p>
</details>

### Access the created EKS Cluster

You can gain access to the Amazon EKS cluster via the `AWS CLI` using the following command:

```shell
aws eks --region "$REGION" update-kubeconfig --name "$CLUSTER_NAME" --alias "$CLUSTER_NAME"
```

After updating the kubeconfig, you can verify your connection to the cluster with kubectl:

```shell
kubectl get nodes
```

Then ensure that a namespace exists for camunda.
In the following procedure, we reference the `camunda` namespace to create some required resources in the Kubernetes cluster, such as secrets or one-time setup jobs.

```shell
kubectl create namespace camunda
```

### Check Existing StorageClasses

To see the available **StorageClasses** in your Kubernetes cluster, use the following command:

```bash
kubectl describe storageclass
```

> This command will display a list of defined StorageClasses, including which one is set as default.

Reviewer: I think we can drop that sentence and just combine it with the one above.

> To see the available and default StorageClasses in your Kubernetes cluster, use the following command:

#### StorageClass

To check if `gp3` is set as the default StorageClass, you can look for the annotation `storageclass.kubernetes.io/is-default-class: "true"` in the output of the previous command. If `gp3` is not installed or not set as default StorageClass, you need to install it and set it as default.

We recommend using **gp3** volumes with Camunda 8 (see [volume performance](./amazon-eks.md#volume-performance)).
It may be necessary to create the StorageClass, as the default configuration only includes **gp2**. For detailed information, refer to the [AWS documentation](https://aws.amazon.com/ebs/general-purpose/).

> #### Create `gp3` StorageClass

Reviewer: should we include a node here that it requires the ebs csi cluster addon?
Just in case someone modified the defaults or is using their own yaml following this guide.

After executing these commands, you will have a `gp3` StorageClass set as the default and the `gp2` StorageClass marked as non-default, provided that **gp2** was already present. Ensure to verify the changes again by running the `kubectl get storageclass` command.

### Requirements for a domain deployment

If you plan to deploy Camunda using an external domain associated with an external certificate, you will need to set up some IAM policies to allow both **external-dns** and **cert-manager** to interact with Route 53, which controls the DNS.

By default, the cluster uses **Pod Identity** to manage IAM roles for your applications. This means that service accounts are associated with IAM roles, allowing your pods to securely access AWS resources without hardcoding credentials. For more information on configuring Pod Identity, refer to the [AWS documentation](https://docs.aws.amazon.com/eks/latest/userguide/pod-identities.html).

Additionally, to [enable OpenID Connect (OIDC) and IAM Roles for Service Accounts (IRSA)](https://docs.aws.amazon.com/eks/latest/userguide/enable-iam-roles-for-service-accounts.html) on your cluster, you can use the following commands:

1. **OIDC Issuer:**

>    First, ensure that your EKS cluster is set up with an OIDC provider. The following command should show you the OIDC issuer:

Reviewer: the phrasing is a bit off.

> First, ensure that your EKS cluster is set up with an OIDC provider.

followed by step 2

> Create an OIDC provider

--- 

From https://docs.aws.amazon.com/eks/latest/userguide/enable-iam-roles-for-service-accounts.html

you first retrieve the OIDC issuer ID, optionally check whether it exists, then create provider.

Alternatively, you can deploy the Helm chart first and then use `eksctl` with the option `--override-existing-serviceaccounts` instead of `--role-only` to reconfigure the created service account.

## 3. PostgreSQL database

Creating a PostgreSQL database can be accomplished through various methods, such as using the AWS Management Console or the AWS CLI. This guide focuses on providing a reproducible setup using the CLI. For information on creating PostgreSQL using the UI, refer to the [AWS documentation](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/CHAP_GettingStarted.CreatingConnecting.PostgreSQL.html).

:::info Optional service

If you don't want to use the Amazon RDS Aurora managed service for PostgreSQL, you can skip this section.
However, please note that you may need to adjust the following instructions to remove references to it.

If you choose not to use this service, you'll need to either provide a managed PostgreSQL service or use the internal deployment by the Camunda Helm chart in Kubernetes.

:::

The following components use the PostgreSQL database:

> - keycloak
> - identity
> - webmodeler

Reviewer: ```suggestion
- Keycloak
- Identity
- Web Modeler
```

       --availability-zones $(echo $ZONES) \
       --db-subnet-group-name camunda-postgres
   ```

   More configuration options can be found in the [AWS documentation](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/rds/create-db-cluster.html).

7. **Wait for the PostgreSQL cluster to be ready:**

   ```shell
   aws rds wait db-cluster-available \
       --db-cluster-identifier $RDS_NAME
   ```

8. **Create a database instance within the DB cluster:**

   Ensure that the `engine-version` matches the previously created PostgreSQL cluster.

   ```shell
   aws rds create-db-instance \
       --db-instance-identifier $RDS_NAME \
       --db-cluster-identifier $RDS_NAME \
       --engine aurora-postgresql \
       --engine-version $POSTGRESQL_VERSION \
       --no-publicly-accessible \
       --db-instance-class db.t3.medium
   ```

   More configuration options can be found in the [AWS documentation](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/rds/create-db-instance.html).

9. **Wait for changes to be applied:**

```shell
aws rds wait db-instance-available \
    --db-instance-identifier $RDS_NAME

> ```

Reviewer: Your indentation is slightly off.

![image](https://github.com/user-attachments/assets/25639b52-a6d4-4e24-8108-738fee96c223)

```

Once the job shows as `Completed`, the users and databases will have been successfully created.

5. **Check logs for confirmation**: You can view the logs of the job to confirm that the users were created and privileges were granted successfully:

```bash
kubectl logs job/create-setup-user-db --namespace camunda
```

6. **Cleanup the resources:**

```bash
kubectl delete job create-setup-user-db --namespace camunda
kubectl delete secret setup-db-secret --namespace camunda
```

By running these commands, you will clean up both the job and the secret, ensuring that no unnecessary resources remain in the cluster.

## 4. OpenSearch domain

Creating an OpenSearch domain can be accomplished through various methods, such as using the AWS Management Console or the AWS CLI. This guide focuses on providing a reproducible setup using the CLI. For information on creating an OpenSearch domain using the UI, refer to the [AWS OpenSearch documentation](https://docs.aws.amazon.com/opensearch-service/latest/developerguide/create-managed-domain.html).

The resulting OpenSearch domain is intended for use with the Camunda platform, the following components utilize OpenSearch:

> - Zeebe
> - Operate
> - Tasklist
> - Optimize

Reviewer: ```suggestion
- Operate
- Optimize
- Tasklist
- Zeebe
```

:::info Optional service

If you don't want to use the Amazon OpenSearch managed service for OpenSearch, you can skip this section.
However, please note that you may need to adjust the following instructions to remove references to it.

If you choose not to use this service, you'll need to either provide a managed OpenSearch or Elasticsearch service or use the internal deployment by the Camunda Helm chart in Kubernetes.

:::

:::note Available since Camunda 8.4

As of the 8.4 release, Zeebe, Operate, and Tasklist are now compatible with [Amazon OpenSearch Service](https://aws.amazon.com/de/opensearch-service/) 2.5.x. Note that using Amazon OpenSearch Service requires [setting up a new Camunda installation](/self-managed/setup/overview.md). A migration from previous versions or Elasticsearch environments is currently not supported.

> :::

Reviewer: Does this note still make sense, considering we're listing couples lines above the ones that work with OpenSearch / Elasticsearch?

Reviewer: Maybe the note could just be generalised to something along the lines of

> It is not feasible to interchangeably switch between Elasticsearch and OpenSearch post Camunda deployment.

   - **Access Policies**: The default access policy allows all actions (`es:*`) on resources within the domain for any AWS account (`"Principal": { "AWS": "*" }`). This is scoped to the OpenSearch domain resources using the `arn:aws:es:$REGION:*:domain/$OPENSEARCH_NAME/*` resource ARN.
   - **VPC Options**: The domain is deployed within the specified VPC, restricted to the provided subnets (`SubnetIds=${SUBNET_IDS}`) and associated security group (`SecurityGroupIds=${GROUP_ID_OPENSEARCH}`).

This configuration creates a secure OpenSearch domain with encryption both in transit (between nodes) and at rest, zonal fault tolerance, and sufficient storage performance using `gp3` volumes. The access is restricted to resources in the VPC of the EKS cluster and is governed by the specified security group.

1. Wait for the OpenSearch domain to be active:

   ```shell
   while [ "$(aws opensearch describe-domain --domain-name $OPENSEARCH_NAME --query 'DomainStatus.Processing' --output text)" != "False" ]; do echo "Waiting for OpenSearch domain to become availablen this can up to take 20-30 minutes..."; sleep 30; done && echo "OpenSearch domain is now available\!"
   ```

> 1. Retrieve the endpoint of the OpenSearch domain:

Reviewer 3: Anything labeled 1. with the proper indentation will be automatically converted to the correct number, so you can move items/add numbers without worrying about impacting the order. It also lets you know when your indentation for numbered lists is incorrect, which can be a problem for accessibility - that's what happened here. 

The indentation should be normalized now! :)
